### DISCUSSION TITLE
remote: Write access to repository not granted.

### DISCUSSION THREAD

**Discussion by tristanrose** [community/community#46398] [Repositories] [answered]
### Select Topic Area

Product Feedback

### Body

hi there,

i am getting a terminal error saying:

```
remote: Write access to repository not granted.
fatal: unable to access 'https://github.com/@username/repo.git/': The requested URL returned error: 403
```

If I am the owner of the repo, why do I not have write access? I created a fine-grained token for this repo but still, nothing.

suggestions from those who solved ran into and solved this before? thanks

**Comment by chris-c-thomas** [comment]
Is that the actual error returned or did you edit it slightly to remove info?

In either case it's likely trying to write to the repository either as a different configured user or no configured user at all. Try running `git config --list` and see what's returned. Hopefully should match the owner account of the repo.

Also, was this the process you took when cloning to use the token? [GitHub Docs: Using a token on the command line](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#using-a-token-on-the-command-line)

**Comment by tristanrose** [comment]
@chris-c-thomas  yep, edited url.  but doubled checked url is the exact match to `git remote add origin <URL>`

just ran `git config --list`,   name and email are synced correct.

but unfortunately, no.  i'm not even getting to the point where i can enter my user and pass (token).  this err is happening before.

i am getting this err as soon as i enter `git push -u origin main`

**Comment by chris-c-thomas** [comment, marked as answer]
Right, you won't be able to push anything until things are configured to use your token instead of your old password which is likely what's happening. That's why I had asked if when you originally cloned the repository you entered your token like this [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#using-a-token-on-the-command-line)?

You can update your cached credentials to your token by following [this doc](https://docs.github.com/en/get-started/getting-started-with-git/updating-credentials-from-the-macos-keychain). There's a link in there about changing to the [Git Credential Manager](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git#git-credential-manager) if you prefer something like that.

You can use the [GitHub CLI](https://cli.github.com/) as well. Running `gh auth login` will let you setup your credentials using your token instead of your old password.

Give these approaches a shot and let me know how it goes.

**Reply by tristanrose** [reply]
brilliant man thanks,  clearing cache following [this doc](https://docs.github.com/en/get-started/getting-started-with-git/updating-credentials-from-the-macos-keychain) did the trick :)

**Reply by chris-c-thomas** [reply]
Awesome! Glad we got it working.

**Reply by sayedsadat98** [reply]
Awesome! Thanks for the guide! Cheers!

**Reply by Sadiq-Rashid** [reply]
The "gh auth login" command helps you set up your credentials using token instead of traditional your password.
The guide was helpful, Cheers!

**Reply by broomer** [reply]
I just needed this as well - thank you!

**Reply by abdramsyah** [reply]
thank for your information, i got same problem and now i can access normal again

**Reply by AnandPatel404** [reply]
thanks a lot this following steps is work for me. 👍🏻  :)

**Comment by kitfbgh** [comment]
Hi guys, I have the same problem but in a different context.
I use the Personal Access Token (Classic) in Travis CI to push tags, and I can push tags normally on January 16, 2023 But then came the 403 error now

suggestions from those who solved ran into and solved this before? thanks

**Reply by chris-c-thomas** [reply]
Well it's likely to be along the same lines. Most likely your password is cached to your `user.email` and your token isn't being used instead. Thus, the 403.

I'd recommend one these two approaches:
1. You can update your credentials in the keychain by following [this guide](https://docs.github.com/en/get-started/getting-started-with-git/updating-credentials-from-the-macos-keychain)
2. You can cache your GitHub credentials using the GitHub CLI or Git Credential Manager following [this guide](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git)

All in all, both of those come from this [main article](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) about Personal Access Tokens in general. You'll want to follow them carefully so your config is set to use your token for the repos that require it.

**Reply by kitfbgh** [reply]
I tried, it didn't help me.
I'm in a CI environment.
Every establishment comes out of image....

**Reply by kesselly4099** [reply]
the reason is sample try to logout using this command $ gh auth logout and go back to GitHub create a new token do forget to give the token the permission when generate, select all the permissions you want the token to have before generating , the go back and to the command  line  and login in using $gh auth login

**Reply by Chaisenberg9k** [reply]
This was it. Thank you. Everyone over complicated everything

**Reply by yagohdezg** [reply]
If someone else is having this problem in CI pipelines, I solved it by adding 
```
permissions:
  contents: write
jobs:
...
```

At the top level of my workflow yaml.

**Reply by mrcentimetre** [reply]
> If someone else is having this problem in CI pipelines, I solved it by adding
> 
> ```
> permissions:
>   contents: write
> jobs:
> ...
> ```
> 
> At the top level of my workflow yaml.

Thanks ;)

**Comment by bobeymcool** [comment]
Finally using the Credential Manager [Git Credential Manager](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git#git-credential-manager) worked for me..

**Reply by indian0ch** [reply]
You just installed it?

**Comment by dyagzy** [comment]
I 'm not a vim or bash savvy developer more especially because I am new to programming. I have a work git account which is associated with my work aws account. With this account, I can push and pull from my work repository. Now I also have a personal git account which I use for personal development. Over the last 1 years or more I haven't used my personal git account but my work account. Recently I decided to fork a private repo with my personal account and got 403 error. I knew instantly that the problem is due to "multiple hit account on same machine". I have applied  all the steps in the [Git Credential Manager](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git#git-credential-manager) , with this I was able to fork this private repo with my personal account and then I couldn't pull or push to my work git repo any more, I had to revert all the changes. My question is how do I switch between different git account without going through the complicated SSH guides which I have been seeing around. I have tried the Git Credential Manger as stated above , I have also tried setting up my intelliji IDE but with this I can only use one account per time and can't switch

**Comment by alexurassa** [comment]
This happened because credentials have changed in remote but not updated in your local git credential store.

So you may use [Git Credential Manager (GCM)](https://docs.github.com/en/get-started/getting-started-with-git/caching-your-github-credentials-in-git#git-credential-manager) to setup so that new credentials from remote will be synced with your local credentials.

Basically there you will follow these two main steps:
   - Install Git Credential Manager (GCM) if you haven't already.
   - Configure Git to use GCM

**Reply by Duhduh420** [reply]
Csyny

**Comment by kesselly4099** [comment]
the reason is sample try to logout using this command $ gh auth logout and go back to GitHub create a new token do forget to give the token the permission when generate, select all the permissions you want the token to have before generating , the go back and to the command line and login in using $gh auth login

**Comment by Duhduh420** [comment]
git remote add origin https://github.com/Duhduh420/Duhduh420.git
git branch -M main
git push -u origin main

**Comment by gonzalezivan90** [comment]
The new key generated here (https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#using-a-personal-access-token-on-the-command-line) must have "writing" permissions. Be sure to click on those options when create the new key. That's how it works for me

**Comment by kerryj89** [comment]
If you landed here because of `fatal: unable to access 'https://github.com/<your-company>/<your-repo>.git/'` or similar issue in GitHub Actions suddenly, I found that I had to explicitly state the permissions in the yaml workflow file now as described above.

I don't know if some change propagated from GitHub or organizational level and I just missed the news, but today my GitHub Actions workflow was failing with the above error.  My workflow pushes tags back to the repo through our own GitHub App.  I reinstalled the app to the organization and to my repository and triple checked my App permissions.  Unless I missed some new permission, everything was set as it was days ago.

HOWEVER... reading through the raw GitHub workflow logs drew a different picture:

Working 2 days ago:
```
2023-10-18T22:05:43.0996961Z ##[group]GITHUB_TOKEN Permissions
2023-10-18T22:05:43.0999064Z Actions: write
2023-10-18T22:05:43.0999735Z Checks: write
2023-10-18T22:05:43.1000452Z Contents: write
2023-10-18T22:05:43.1001194Z Deployments: write
2023-10-18T22:05:43.1001714Z Discussions: write
2023-10-18T22:05:43.1002296Z Issues: write
2023-10-18T22:05:43.1002855Z Metadata: read
2023-10-18T22:05:43.1003338Z Packages: write
2023-10-18T22:05:43.1003919Z Pages: write
2023-10-18T22:05:43.1004406Z PullRequests: write
2023-10-18T22:05:43.1004999Z RepositoryProjects: write
2023-10-18T22:05:43.1005892Z SecurityEvents: write
2023-10-18T22:05:43.1006436Z Statuses: write
2023-10-18T22:05:43.1006972Z ##[endgroup]
```

Error today:
```
2023-10-20T21:00:35.4030454Z ##[group]GITHUB_TOKEN Permissions
2023-10-20T21:00:35.4032814Z Contents: read
2023-10-20T21:00:35.4033545Z Metadata: read
2023-10-20T21:00:35.4034285Z Packages: read
2023-10-20T21:00:35.4034945Z ##[endgroup]
```

It seems GitHub (or my organization) is preventing GitHub App permissions from casting its wide permission net implicitly to my workflow.  After adding back the missing permissions from 2 days ago into my workflow yaml, it started working again.

Edit: Leaving this up since it may still be helpful, but I just learned that our org upgraded to GitHub Enterprise which I guess affected workflows.

**Reply by serban-seeq** [reply]
I believe we hit the same issue when I was trying to run  `git push origin <branchname>`  as part of an action step. The odd thing is it works for all branches except `main`

**Comment by Duhduh420** [comment]
Lol

**Comment by Aristarque04** [comment]
not get any results guys any other way?

**Comment by Hamed-Hasan** [comment]
It seems like you're encountering a 403 error, indicating that there's an issue with write access to your GitHub repository. This could be related to your authentication token or repository settings. Let's troubleshoot this:

1. **Check Token Permissions:**
   Ensure that the personal access token you generated has the necessary permissions. Make sure it has "repo" or "public_repo" scope if your repository is public, and "repo" scope for private repositories.

2. **Token Usage:**
   Confirm that you are using the correct token. Sometimes, users accidentally use their password instead of the access token. Update your Git credentials to use the correct token.

   ```bash
   git remote set-url origin https://<token>@github.com/username/repo.git
   ```

3. **Verify Repository Access:**
   Double-check the repository settings to ensure that you have the appropriate access level (write access) as the owner. Go to your repository on GitHub, navigate to "Settings," then "Manage access" to confirm your access level.

4. **Token Regeneration:**
   If the issue persists, try regenerating a new access token and update your Git credentials with the new token.

5. **HTTPS vs. SSH:**
   Ensure that you are using the correct URL. If you are using HTTPS, make sure the remote URL is correct. If you are using SSH, confirm that your SSH key is correctly set up.

**Reply by DILNATHRK** [reply]
git remote set-url origin https://<token>@github.com/username/repo.git                                                                                                                 this worked for me , Thankyou

**Comment by de-jagaban** [comment]
Well, i experienced the same issue and to fix this i simply went back to Git PAT i created and updated permissions. The issue was that i had selected a list of my repos that should have access to that particular PAT (I totally forgot that), and my new repo was not included, so i simply edited the PAT and added the new repo and it worked.

**Comment by EvanCarroll** [comment]
I had a totally different problem causing this,

https://devops.stackexchange.com/q/19922/18965

**Comment by hechmi10** [comment]
remote: Write access to repository not granted.
fatal: unable to access 'https://github.com/Pfa-EduBridge/Backend_microservice/': The requested URL returned error: 403
It's in another branch what can I do?

**Comment by Tectract** [comment]
Someone just gave me read access to their repo, how do I clone the repo in my github codespace Jupyter terminal? I'm suspicious this is a malicious repo so there's no way I'm running it locally, lol.

Anyone?

Cloning into 'spacexview_v1'...
remote: Write access to repository not granted.
fatal: unable to access 'https://github.com/UnicsoftOrgs/spacexview_v1/': The requested URL returned error: 403

What do I need to do to clone repos I should have read access to? This should not be so insanely difficult.
